File: Tests/3-swift/Sources/FJ/V2/FJInterpreter.swift
// Based on <https://github.com/fjpub/fj-lam/blob/cbcb453fb2595d341bb3c1341fdf235f066778fd/V2/FJInterpreter.hs>.

/// Evaluates an expression.
/// - Returns: An expression after processing one reduction step.
func evalʹ(ct classTable: CT, expr: FJExpr) -> FJExpr? {
  switch expr {
  case let .createObject(className, params): // RC-New-Arg
    let pʹ = params.compactMap { evalʹ(ct: classTable, expr: $0) }
    return .createObject(className, pʹ)

  case .fieldAccess(let expr, let fieldName):
    if isValue(ct: classTable, expr) { // R-Field
      switch expr {
      case let .createObject(className, params):
        guard let fields = fields(ct: classTable, className: className) else {
          assertionFailure()
          return nil
        }
        guard let index = fields.firstIndex(where: { $0.name == fieldName }) else {
          assertionFailure()
          return nil
        }
        return lambdaMark(expression: params[index], type: fields[index].name)
      default:
        return nil // Not an object instance
      }
    } else { // RC-Field
      return evalʹ(ct: classTable, expr: expr).map { .fieldAccess($0, fieldName) }
    }

  case let .methodInvocation(source, methodName, methodParams):
    if isValue(ct: classTable, source) {
      if methodParams.allSatisfy({ isValue(ct: classTable, $0) }) {
        // R-Invk
        switch source {
        case let .createObject(className, _): // R-Invk
          guard let (methodParameterNames, methodBody) = methodBody(
            ct: classTable,
            methodName: methodName,
            className: className
          ) else { return nil } // No method body
          guard let (methodArgTypes, methodReturnType) = methodType(
            ct: classTable,
            methodName: methodName,
            className: className
          ) else { return nil } // No method type
          let pʹ = zip(methodParams, methodArgTypes).map(lambdaMark)
          return substitute(
            parameterNames: methodParameterNames + ["this"],
            parameters: pʹ + [source],
            body: lambdaMark(expression: methodBody, type: methodReturnType)
          )
        case let .cast(className, .lambda(lambdaParams, lambdaBody)):
          guard let (methodArgTypes, methodReturnType) = methodType(
            ct: classTable,
            methodName: methodName,
            className: className
          ) else { return nil } // No method type
          let pʹ = zip(methodParams, methodArgTypes).map(lambdaMark)

          if let (methodParameterNames, methodBody) = methodBody(
            ct: classTable,
            methodName: methodName,
            className: className
          ) { // R-Default
            return substitute(
              parameterNames: methodParameterNames,
              parameters: pʹ,
              body: lambdaMark(expression: methodBody, type: methodReturnType)
            )
          } else { // R-Lam
            return substitute(
              parameterNames: lambdaParams.map(\.name),
              parameters: pʹ,
              body: lambdaMark(expression: lambdaBody, type: methodReturnType)
            )
          }
        default:
          return nil
        }
      } else { // RC-Invk-Arg
        let pʹ = methodParams.compactMap { evalʹ(ct: classTable, expr: $0) }
        return .methodInvocation(source, methodName, pʹ)
      }
    } else { // RC-Invk-Recv
      return evalʹ(ct: classTable, expr: source).map {
        .methodInvocation($0, methodName, methodParams)
      }
    }

  case let .cast(castType, castExpr):
    if isValue(ct: classTable, castExpr) {
      switch castExpr {
      case let .createObject(type, _):
        if subtyping(ct: classTable, type, castType) { // R-Cast
          return castExpr
        } else {
          return nil
        }
      case let .cast(lambdaType, .lambda):
        if subtyping(ct: classTable, lambdaType, castType) { // R-Cast-Lam
          return castExpr
        } else {
          return nil
        }
      default:
        return expr // Annotated lambda expression is a value
      }
    } else { // RC-Cast
      return evalʹ(ct: classTable, expr: castExpr).map { FJExpr.cast(castType, $0) }
    }

  case .lambda:
    return expr

  case .variable:
    return nil
  }
}

/// Evaluates an expression recursively.
/// - Returns: A value after all the reduction steps.
func eval(ct classTable: CT, expr: FJExpr) -> FJExpr {
  if isValue(ct: classTable, expr) {
    return expr
  } else {
    return eval(ct: classTable, expr: evalʹ(ct: classTable, expr: expr) ?? expr)
  }
}

/// Replaces actual parameters in method body expression.
/// - Returns: A new changed expression.
func substitute(parameterNames: [String], parameters: [FJExpr], body: FJExpr) -> FJExpr? {
  switch body {
  case .variable(let varName):
    return parameterNames.firstIndex(of: varName).map { parameters[$0] }
  case let .fieldAccess(source, fieldName):
    return substitute(parameterNames: parameterNames, parameters: parameters, body: source)
      .map { FJExpr.fieldAccess($0, fieldName) }
  case let .methodInvocation(source, methodName, methodParams):
    let methodParams = methodParams.compactMap {
      substitute(parameterNames: parameterNames, parameters: parameters, body: $0)
    }
    return substitute(parameterNames: parameterNames, parameters: parameters, body: source)
      .map { FJExpr.methodInvocation($0, methodName, methodParams) }
  case let .createObject(className, parameters):
    let parameters = parameters.compactMap {
      substitute(parameterNames: parameterNames, parameters: parameters, body: $0)
    }
    return .createObject(className, parameters)
  case let .cast(castType, expr):
    return substitute(parameterNames: parameterNames, parameters: parameters, body: expr)
      .map { FJExpr.cast(castType, $0) }
  case .lambda:
    return body // Do nothing
  }
}

File: Explorations/4-java/src/imt.fil.a3/recherche.fj/v2/FJInterpreter.java
package imt.fil.a3.recherche.fj.v2;

import imt.fil.a3.recherche.fj.FJUtils;
import imt.fil.a3.recherche.fj.parser.FJField;
import imt.fil.a3.recherche.fj.parser.FJMethodBodySignature;
import imt.fil.a3.recherche.fj.parser.FJMethodTypeSignature;
import imt.fil.a3.recherche.fj.parser.expression.*;
import imt.fil.a3.recherche.fj.parser.type.FJType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

public record FJInterpreter(HashMap<String, FJType> classTable) {

    /**
     * Evaluates an expression.
     *
     * @param expr An expression.
     * @return An expression after processing one reduction step.
     */
    private Optional<FJExpr> _eval(FJExpr expr) {
        if (expr instanceof final FJCreateObject createObject) { // RC-New-Arg
            final List<FJExpr> args = createObject.args().stream().map(this::_eval)
                .flatMap(Optional::stream).toList();
            return Optional.of(new FJCreateObject(createObject.className(), args));
        } else if (expr instanceof final FJFieldAccess fieldAccess) { // R-Field
            if (fieldAccess.object().isValue()) {
                if (fieldAccess.object() instanceof final FJCreateObject createObject) {
                    final Optional<List<FJField>> _fields =
                        FJUtils.classFields(this.classTable, createObject.className());
                    if (_fields.isEmpty()) return Optional.empty();
                    final List<FJField> fields = _fields.get();

                    Optional<Integer> index = Optional.empty();
                    for (int i = 0; i < fields.size(); i++) {
                        if (fields.get(i).name().equals(fieldAccess.fieldName())) {
                            index = Optional.of(i);
                            break;
                        }
                    }
                    if (index.isEmpty()) return Optional.empty();

                    final FJExpr arg = createObject.args().get(index.get());
                    final String lambdaTypeName = fields.get(index.get()).name();
                    return Optional.of(arg.lambdaMark(lambdaTypeName));
                } else {
                    return Optional.empty(); // Not an object instance
                }
            } else { // RC-Field
                return this._eval(expr).map(e -> new FJFieldAccess(e, fieldAccess.fieldName()));
            }
        } else if (expr instanceof final FJMethodInvocation methodInvocation) {
            if (methodInvocation.source().isValue()) {
                if (methodInvocation.args().stream().allMatch(FJExpr::isValue)) {
                    if (methodInvocation.source() instanceof final FJCreateObject createObject) { // R-Invk
                        final Optional<FJMethodTypeSignature> methodType = FJUtils.methodType(
                            this.classTable,
                            methodInvocation.methodName(),
                            createObject.className()
                        );
                        if (methodType.isEmpty()) return Optional.empty(); // No method type

                        final Optional<FJMethodBodySignature> methodBody = FJUtils.methodBody(
                            this.classTable,
                            methodInvocation.methodName(),
                            createObject.className()
                        );
                        if (methodBody.isEmpty()) return Optional.empty(); // No method body

                        final List<FJExpr> args = new ArrayList<>();
                        // <=> zip(methodInvocation.args, methodType.parameterTypeNames)
                        for (int i = 0; i < methodInvocation.args().size(); i++) {
                            final FJExpr arg = methodInvocation.args().get(i);
                            final String typeName = methodType.get().parameterTypeNames().get(i);
                            args.add(arg.lambdaMark(typeName));
                        }
                        args.add(methodInvocation.source());
                        final List<String> parameterNames = Stream.concat(
                            methodBody.get().argumentNames().stream(),
                            Stream.of("this")
                        ).toList();
                        return this.substitute(
                            parameterNames,
                            args,
                            methodBody.get().body().lambdaMark(methodType.get().returnTypeName())
                        );
                    } else if (
                        methodInvocation.source() instanceof final FJCast fjCast
                            && fjCast.body() instanceof final FJLambda lambda
                    ) {
                        final Optional<FJMethodTypeSignature> methodType = FJUtils.methodType(
                            this.classTable,
                            methodInvocation.methodName(),
                            fjCast.typeName()
                        );
                        if (methodType.isEmpty()) return Optional.empty(); // No method type
                        final List<FJExpr> args = new ArrayList<>();
                        // <=> zip(methodInvocation.args, methodType.parameterTypeNames)
                        for (int i = 0; i < methodInvocation.args().size(); i++) {
                            final FJExpr arg = methodInvocation.args().get(i);
                            final String typeName = methodType.get().parameterTypeNames().get(i);
                            args.add(arg.lambdaMark(typeName));
                        }

                        final Optional<FJMethodBodySignature> methodBody = FJUtils.methodBody(
                            this.classTable,
                            methodInvocation.methodName(),
                            fjCast.typeName()
                        );
                        if (methodBody.isPresent()) { // R-Default
                            return substitute(
                                methodBody.get().argumentNames(),
                                args,
                                methodBody.get().body().lambdaMark(methodType.get().returnTypeName())
                            );
                        } else { // R-Lam
                            return this.substitute(
                                lambda.args().stream().map(FJField::name).toList(),
                                args,
                                lambda.body().lambdaMark(methodType.get().returnTypeName())
                            );
                        }
                    } else {
                        return Optional.empty();
                    }
                } else { // RC-Invk-Arg
                    final List<FJExpr> args = methodInvocation.args().stream().map(this::_eval)
                        .flatMap(Optional::stream).toList();
                    return Optional.of(new FJMethodInvocation(
                        methodInvocation.source(),
                        methodInvocation.methodName(),
                        args
                    ));
                }
            } else { // RC-Invk-Recv
                return this._eval(methodInvocation.source())
                    .map(e -> new FJMethodInvocation(e, methodInvocation.methodName(), methodInvocation.args()));
            }
        } else if (expr instanceof final FJCast fjCast) {
            if (fjCast.body().isValue()) {
                if (fjCast.body() instanceof final FJCreateObject createObject) {
                    if (FJUtils.isSubtype(this.classTable, createObject.className(), fjCast.typeName())) { // R-Cast
                        return Optional.of(fjCast.body());
                    } else {
                        return Optional.empty();
                    }
                } else if (fjCast.body() instanceof final FJCast fjCast2 && fjCast2.body() instanceof FJLambda) {
                    if (FJUtils.isSubtype(this.classTable, fjCast2.typeName(), fjCast.typeName())) { // R-Cast-Lam
                        return Optional.of(fjCast.body());
                    } else {
                        return Optional.empty();
                    }
                } else {
                    return Optional.empty();
                }
            } else { // RC-Cast
                return this._eval(fjCast.body()).map(e -> new FJCast(fjCast.typeName(), e));
            }
        } else if (expr instanceof FJLambda) {
            return Optional.of(expr);
        } else if (expr instanceof FJVariable) {
            return Optional.empty();
        } else {
            throw new RuntimeException("Unexpected code path: expression type not supported.");
        }
    }

    /**
     * Evaluates an expression recursively.
     *
     * @return A value after all the reduction steps.
     */
    public FJExpr eval(FJExpr expr) {
        if (expr.isValue()) {
            return expr;
        } else {
            return eval(_eval(expr).orElse(expr));
        }
    }

    /**
     * Replaces actual parameters in method body expression.
     *
     * @return A new changed expression.
     */
    private Optional<FJExpr> substitute(List<String> parameterNames, List<FJExpr> args, FJExpr body) {
        if (body instanceof final FJVariable fjVariable) {
            final int index = parameterNames.indexOf(fjVariable.name());
            if (index >= 0 && args.size() > index) {
                return Optional.of(args.get(index));
            } else {
                return Optional.empty();
            }
        } else if (body instanceof final FJFieldAccess fieldAccess) {
            return this.substitute(parameterNames, args, fieldAccess.object())
                .map(e -> new FJFieldAccess(e, fieldAccess.fieldName()));
        } else if (body instanceof final FJMethodInvocation methodInvocation) {
            return this.substitute(parameterNames, args, methodInvocation.source())
                .map(e -> {
                    final List<FJExpr> _args = methodInvocation.args().stream()
                        .map(a -> this.substitute(parameterNames, args, a))
                        .flatMap(Optional::stream).toList();
                    return new FJMethodInvocation(e, methodInvocation.methodName(), _args);
                });
        } else if (body instanceof final FJCreateObject createObject) {
            final List<FJExpr> _args = createObject.args().stream()
                .map(a -> this.substitute(parameterNames, args, a))
                .flatMap(Optional::stream).toList();
            return Optional.of(new FJCreateObject(createObject.className(), _args));
        } else if (body instanceof final FJCast fjCast) {
            return this.substitute(parameterNames, args, fjCast.body())
                .map(e -> new FJCast(fjCast.typeName(), e));
        } else if (body instanceof FJLambda) {
            return Optional.of(body); // Do nothing
        } else {
            throw new RuntimeException("Unexpected code path: expression type not supported.");
        }
    }

}
